File: modern/src/com/cattailsw/nanidroid/ComposeBackedGhostPresentationRenderer.java
package com.cattailsw.nanidroid;

import android.view.View;

import com.cattailsw.nanidroid.compose.GhostPresentationComposeHost;

/**
 * Keeps the legacy surface engine authoritative while Compose renders balloons.
 */
public final class ComposeBackedGhostPresentationRenderer implements GhostPresentationRenderer {
    private final LegacyGhostPresentationRenderer surfaceRenderer;
    private final Balloon sakuraBalloon;
    private final Balloon keroBalloon;
    private final GhostPresentationComposeHost composeHost;

    public ComposeBackedGhostPresentationRenderer(
            SakuraView sakura,
            KeroView kero,
            Balloon sakuraBalloon,
            Balloon keroBalloon,
            LayoutManager layoutManager,
            GhostPresentationComposeHost composeHost) {
        surfaceRenderer = new LegacyGhostPresentationRenderer(
                sakura, kero, sakuraBalloon, keroBalloon, layoutManager);
        this.sakuraBalloon = sakuraBalloon;
        this.keroBalloon = keroBalloon;
        this.composeHost = composeHost;
    }

    @Override
    public void render(GhostPresentationFrame frame) {
        surfaceRenderer.render(frame);
        boolean sakuraUsesLegacyInteraction = requiresLegacyInteraction(sakuraBalloon);
        boolean keroUsesLegacyInteraction = requiresLegacyInteraction(keroBalloon);
        if (!sakuraUsesLegacyInteraction) {
            sakuraBalloon.setVisibility(View.INVISIBLE);
        }
        if (!keroUsesLegacyInteraction) {
            keroBalloon.setVisibility(View.INVISIBLE);
        }
        composeHost.render(frame, !sakuraUsesLegacyInteraction, !keroUsesLegacyInteraction);
    }

    /**
     * Compose owns static presentation only.  The retained TextView remains
     * authoritative when it has a link or overflow scrolling behavior that
     * Compose has not migrated yet.
     */
    private static boolean requiresLegacyInteraction(Balloon balloon) {
        return balloon.getVisibility() == View.VISIBLE
                && (balloon.getUrls().length != 0 || balloon.getMovementMethod() != null);
    }
}

File: modern/src/com/cattailsw/nanidroid/ComposeBackedGhostPresentationRenderer.kt
package com.cattailsw.nanidroid

import android.view.View
import com.cattailsw.nanidroid.compose.GhostPresentationComposeHost

/** Keeps the retained surface engine authoritative while Compose renders balloons. */
class ComposeBackedGhostPresentationRenderer(
    sakura: SakuraView,
    kero: KeroView,
    private val sakuraBalloon: Balloon,
    private val keroBalloon: Balloon,
    layoutManager: LayoutManager?,
    private val composeHost: GhostPresentationComposeHost,
) : GhostPresentationRenderer {
    private val surfaceRenderer = LegacyGhostPresentationRenderer(
        sakura, kero, sakuraBalloon, keroBalloon, layoutManager,
    )

    override fun render(frame: GhostPresentationFrame) {
        surfaceRenderer.render(frame)
        val sakuraUsesLegacyInteraction = requiresLegacyInteraction(sakuraBalloon)
        val keroUsesLegacyInteraction = requiresLegacyInteraction(keroBalloon)
        if (!sakuraUsesLegacyInteraction) sakuraBalloon.visibility = View.INVISIBLE
        if (!keroUsesLegacyInteraction) keroBalloon.visibility = View.INVISIBLE
        composeHost.render(frame, !sakuraUsesLegacyInteraction, !keroUsesLegacyInteraction)
    }

    private fun requiresLegacyInteraction(balloon: Balloon): Boolean =
        balloon.visibility == View.VISIBLE && (balloon.urls.isNotEmpty() || balloon.movementMethod != null)
}
